--- aes/src/rtl/aes_sbox.v	2026-03-13 08:24:38.692668831 +0100
+++ src_rtl/aes_sbox_parallel.v	2026-03-13 09:29:24.829627542 +0100
@@ -1,6 +1,6 @@
 //======================================================================
 //
-// aes_sbox.v
+// aes_sbox_parallel.v
 // ----------
 // The AES S-box. Basically a 256 Byte ROM. This implementation
 // contains four parallel S-boxes to handle a 32 bit word.
@@ -39,9 +39,9 @@
 
 `default_nettype none
 
-module aes_sbox(
-                input wire [31 : 0]  sboxw,
-                output wire [31 : 0] new_sboxw
+module aes_sbox_parallel( // Modified to be 16 parallel S-boxes
+                input wire [127 : 0]  sboxw,
+                output wire [127 : 0] new_sboxw
                );
 
 
@@ -52,8 +52,20 @@
 
 
   //----------------------------------------------------------------
-  // Four parallel muxes.
+  // 16 parallel muxes.
   //----------------------------------------------------------------
+  assign new_sboxw[127:120] = sbox[sboxw[127:120]];
+  assign new_sboxw[119:112] = sbox[sboxw[119:112]];
+  assign new_sboxw[111:104] = sbox[sboxw[111:104]];
+  assign new_sboxw[103: 96] = sbox[sboxw[103: 96]];
+  assign new_sboxw[95 : 88] = sbox[sboxw[95 : 88]];
+  assign new_sboxw[87 : 80] = sbox[sboxw[87 : 80]];
+  assign new_sboxw[79 : 72] = sbox[sboxw[79 : 72]];
+  assign new_sboxw[71 : 64] = sbox[sboxw[71 : 64]];
+  assign new_sboxw[63 : 56] = sbox[sboxw[63 : 56]];
+  assign new_sboxw[55 : 48] = sbox[sboxw[55 : 48]];
+  assign new_sboxw[47 : 40] = sbox[sboxw[47 : 40]];
+  assign new_sboxw[39 : 32] = sbox[sboxw[39 : 32]];
   assign new_sboxw[31 : 24] = sbox[sboxw[31 : 24]];
   assign new_sboxw[23 : 16] = sbox[sboxw[23 : 16]];
   assign new_sboxw[15 : 08] = sbox[sboxw[15 : 08]];
@@ -320,8 +332,8 @@
   assign sbox[8'hfe] = 8'hbb;
   assign sbox[8'hff] = 8'h16;
 
-endmodule // aes_sbox
+endmodule // aes_sbox_parallel
 
 //======================================================================
-// EOF aes_sbox.v
+// EOF aes_sbox_parallel.v
 //======================================================================
